src/device_handle/async_api.rs

```diff
@@ -0,0 +1,266 @@
+            .expect("Duration was too long to fit into a c_uint");
+
+        // Safety: Pinning `result` ensures it doesn't move, but we know that we will
+        // want to access its fields mutably, we just don't want its memory location
+        // changing (or its fields moving!). So routinely we will unsafely interact with
+        // its fields mutably through a shared reference, but this is still sound.
+        let result = Box::pin(Self {
+            ptr,
+            pool_id,
+            buffer,
+            completed_queue,
+            device,
+            _pin: PhantomPinned,
+        });
+        let user_data = std::ptr::addr_of!(*result) as *mut c_void;
+
+        unsafe {
+            // Step 2 of async api
+            ffi::libusb_fill_bulk_transfer(
+                ptr.as_ptr(),
+                result.device.as_raw(),
+                endpoint,
+                result.buffer.as_ptr() as *mut u8,
+                result.buffer.capacity().try_into().unwrap(),
+                Self::transfer_cb,
+                user_data, // We haven't associated with a queue yet, so this is null
+                timeout,
+            )
+        };
+        result
+    }
+
+    // Part of step 4 of async API the transfer is finished being handled when
+    // `poll()` is called.
+    extern "system" fn transfer_cb(transfer: *mut ffi::libusb_transfer) {
+        // Safety: libusb should never make this null, so this is fine
+        let transfer = unsafe { &mut *transfer };
+
+        // sanity
+        debug_assert_eq!(
+            transfer.transfer_type,
+            ffi::constants::LIBUSB_TRANSFER_TYPE_BULK
+        );
+
+        let transfer: *mut AsyncTransfer<C> = transfer.user_data.cast();
+        let transfer = unsafe { &mut *transfer };
+        // Mark transfer as completed
+        transfer
+            .completed_queue
+            .lock()
+            .unwrap()
+            .push_back(transfer.pool_id);
+    }
+
+    // Step 3 of async API
+    fn submit(self: &mut Transfer<C>) -> Result<(), AsyncError> {
+        let transfer_struct = self.ptr;
+        let errno = unsafe { ffi::libusb_submit_transfer(transfer_struct.as_ptr()) };
+
+        use ffi::constants::*;
+        use AsyncError as E;
+        match errno {
+            0 => Ok(()),
+            LIBUSB_ERROR_NO_DEVICE => Err(E::Disconnected),
+            LIBUSB_ERROR_BUSY => {
+                unreachable!("We shouldn't be calling submit on transfers already submitted!")
+            }
+            LIBUSB_ERROR_NOT_SUPPORTED => Err(E::Other("Transfer not supported")),
+            LIBUSB_ERROR_INVALID_PARAM => Err(E::Other("Transfer size bigger than OS supports")),
+            _ => Err(E::Errno("Error while submitting transfer: ", errno)),
+        }
+    }
+}
+impl<C: UsbContext> Drop for AsyncTransfer<C> {
+    fn drop(&mut self) {
+        // TODO: Figure out how to destroy transfers, which is step 5 of async API.
+        todo!()
```

Comment:
My suggestion is that it should cancel any pending transfers, and then block waiting for them to be reaped. That substantially simplifies the memory management if you don't have to worry about things completing after drop.

Cancellation only requires a round trip into the kernel to clean up the transfers, and the timing doesn't depend on device IO. Haven't timed it, but expect to be a few milliseconds at most. It's a similar situation to `Write` impls that flush and close a file descriptor on drop, which seems in line with RAII.

If users need more control over shutdown, they can cancel and/or drain the transfers before drop, and then drop just frees memory.

Reply:
OK, I believe I have implemented this now